testing/integration/otel_test.go

```diff
@@ -1099,3 +1100,191 @@ service:
+    mapping:
+      mode: bodymap
+service:
+  pipelines:
+    logs:
+      receivers:
+        - filebeatreceiver
+      exporters:
+        - elasticsearch/log
+`
+	otelConfigPath := filepath.Join(tmpDir, "otel.yml")
+	var otelConfigBuffer bytes.Buffer
+	require.NoError(t,
+		template.Must(template.New("otelConfig").Parse(otelConfigTemplate)).Execute(&otelConfigBuffer,
+			otelConfigOptions{
+				InputPath:  inputFilePath,
+				HomeDir:    tmpDir,
+				ESEndpoint: esEndpoint,
+				ESApiKey:   esApiKey.Encoded,
+				Index:      index,
+			}))
+	require.NoError(t, os.WriteFile(otelConfigPath, otelConfigBuffer.Bytes(), 0o600))
+	t.Cleanup(func() {
+		if t.Failed() {
+			contents, err := os.ReadFile(otelConfigPath)
+			if err != nil {
+				t.Logf("No otel configuration file at %s", otelConfigPath)
+				return
+			}
+			t.Logf("Contents of otel config file:\n%s\n", string(contents))
+		}
+	})
+	// Now we can actually create the fixture and run it
+	fixture, err := define.NewFixtureFromLocalBuild(t, define.Version(), aTesting.WithAdditionalArgs([]string{"--config", otelConfigPath}))
+	require.NoError(t, err)
+
+	ctx, cancel := testcontext.WithDeadline(t, context.Background(), time.Now().Add(5*time.Minute))
+	defer cancel()
+	err = fixture.Prepare(ctx, fakeComponent)
+	require.NoError(t, err)
+
+	// Write logs to input file
+	var inputLinesCounter atomic.Int64
+	var stopInputWriter atomic.Bool
+	go func() {
+		for i := 0; ; i++ {
+			if stopInputWriter.Load() {
+				break
+			}
+
+			_, err = inputFile.Write([]byte(fmt.Sprintf("Line %d\n", i)))
+			require.NoErrorf(t, err, "failed to write line %d to temp file", i)
+			inputLinesCounter.Add(1)
+			time.Sleep(100 * time.Millisecond)
+		}
+		err = inputFile.Close()
+		require.NoError(t, err, "failed to close input file")
+	}()
+
+	t.Cleanup(func() {
+		if t.Failed() {
+			contents, err := os.ReadFile(inputFilePath)
+			if err != nil {
+				t.Logf("no data file to import at %s", inputFilePath)
+				return
+			}
+			t.Logf("contents of import file:\n%s\n", string(contents))
+		}
+	})
+
+	var restartCount atomic.Int64
+	var fixtureWg sync.WaitGroup
+	fixtureWg.Add(1)
+	go func() {
+		defer fixtureWg.Done()
+
+		for {
+			// restart the collector every couple seconds while new data is being written
+			prevCount := inputLinesCounter.Load()
+			fCtx, cancel := context.WithDeadline(ctx, time.Now().Add(2*time.Second))
+			err = fixture.RunOtelWithClient(fCtx)
+			cancel()
+			require.True(t, errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled), "unexpected error: %v", err)
+			require.True(t, inputLinesCounter.Load() > prevCount, "expected input lines to increase")
+			restartCount.Add(1)
+
+			if stopInputWriter.Load() {
+				break
+			}
+		}
+
+		// start the collector again for the remaining of the test
+		fCtx, cancel := context.WithDeadline(ctx, time.Now().Add(5*time.Minute))
+		defer cancel()
+		err = fixture.RunOtelWithClient(fCtx)
+	}()
+
+	time.Sleep(5 * time.Second)
```

Comment:
This might be better as a counter of the number of times the collector has restarted. Then you only need to maintain a single arbitrary time interval, the delay between restarts.

Reply:
Instead of a require.True that restartCount is greater than zero, you would just delay until it cross a threshold.

Perhaps better ways to do that than using an atomic and busy waiting on it as well: number of writes to a channel, condition variable, etc.